test/opuntia_SUITE.erl

```diff
@@ -233,34 +238,43 @@ success_or_log_and_return(false, S, P) ->
     ct:pal(S, P),
     false.
 
-should_take_in_range(#{rate := Rate, start_full := false}, ToConsume) ->
-    ExpectedMs = ToConsume / Rate,
-    {ExpectedMs, ExpectedMs + 1};
+should_take_in_range(#{rate := Rate, time_unit := TimeUnit, start_full := false}, ToConsume) ->
+    Expected = ToConsume / Rate,
+    ExpectedMs = opuntia:convert_time_unit(Expected, TimeUnit, millisecond),
+    {ExpectedMs, ceil(ExpectedMs + 1)};
 should_take_in_range(#{bucket_size := MaximumTokens,
                        rate := Rate,
+                       time_unit := TimeUnit,
                        start_full := true},
                      ToConsume) ->
     case ToConsume < MaximumTokens of
         true -> {0, 0};
         false ->
             ToThrottle = ToConsume - MaximumTokens,
-            ExpectedMs = ToThrottle / Rate,
-            {ExpectedMs, ExpectedMs + 1}
+            Expected = ToThrottle / Rate,
+            ExpectedMs = opuntia:convert_time_unit(Expected, TimeUnit, millisecond),
+            {ExpectedMs, ceil(ExpectedMs + 1)}
     end.
 
 run_shaper(Shape, ToConsume) ->
-    Now = erlang:monotonic_time(),
-    Shaper = opuntia:create(Shape, Now),
-    run_shaper(Shaper, Now, [], 0, ToConsume).
+    Shaper = opuntia:create(Shape, 0),
+    run_shaper(Shaper, [], 0, ToConsume).
 
-run_shaper(Shaper, _, History, AccumulatedDelay, 0) ->
+run_shaper(Shaper, History, AccumulatedDelay, 0) ->
     {Shaper, lists:reverse(History), AccumulatedDelay};
-run_shaper(Shaper, Now, History, AccumulatedDelay, TokensLeft) ->
+run_shaper(Shaper, History, AccumulatedDelay, TokensLeft) ->
     %% Uniform distributes in [1, N], and we want [0, N], so we generate [1, N+1] and subtract 1
     ConsumeNow = rand:uniform(TokensLeft + 1) - 1,
-    {NewShaper, DelayMs} = opuntia:calculate(Shaper, ConsumeNow, Now),
-    true = DelayMs >= 0,
-    run_shaper(NewShaper, Now, [{ConsumeNow, DelayMs} | History], AccumulatedDelay + DelayMs, TokensLeft - ConsumeNow).
+    {NewShaper, DelayMs} = opuntia:calculate(Shaper, ConsumeNow, 0),
+    NewHistory = [{ConsumeNow, DelayMs, NewShaper} | History],
+    NewDelay = AccumulatedDelay + DelayMs,
+    NewToConsume = TokensLeft - ConsumeNow,
+    case is_integer(DelayMs) andalso DelayMs >= 0 of
```

Comment:
The type of `DelayMs` should be `non_neg_integer()`, so why would you expect anything else?

Reply:
I'm checking in tests that that is actually the case, because the type definition might end up failing, disadvantage of dynamic typing 🥲

Reply:
But it looks like we accept a bad result (not matching the type spec), because the test doesn't fail in the `false` clause.

Reply:
Hmm... but we stop running and will fail later. Actually, I've just noticed here I'm returning a 4-tuple, which will fail with an error. I'll return a value that will always fall outside of the conditioned range instead.